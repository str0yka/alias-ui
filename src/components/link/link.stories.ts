import type { Meta, StoryObj } from '@storybook/react';

import { Link } from './link';

const meta = {
  title: 'Components/Link/Modes',
  component: Link,
  parameters: {
    layout: 'centered'
  },
  argTypes: {
    disabled: {
      control: 'boolean'
    }
  },
  args: {
    disabled: false
  }
} satisfies Meta<typeof Link>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Internal: Story = {
  name: 'Internal',
  args: {
    mode: 'internal',
    children: 'внутренняя ссылка'
  }
};

export const External: Story = {
  name: 'External',
  args: {
    mode: 'external',
    children: 'внешняя ссылка'
  }
};
