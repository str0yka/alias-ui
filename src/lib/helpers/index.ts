export { cn } from './cn';
export { getTestProps } from './tests';
