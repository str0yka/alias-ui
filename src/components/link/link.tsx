import { cva } from 'class-variance-authority';
import { forwardRef } from 'react';

import { cn } from '@/lib';

export type LinkMode = 'internal' | 'external';

export interface LinkProps extends React.ComponentProps<'a'> {
  mode?: LinkMode;
  disabled?: boolean;
}

export const linkStyles = cva(
  [
    'typography-regular-12 text-info-400 cursor-pointer select-none underline-offset-2',
    'hover:text-info-300',
    'active:text-info-300'
  ],
  {
    variants: {
      mode: {
        internal: 'active:underline',
        external: ['underline decoration-dashed', 'active:no-underline']
      }
    }
  }
);

export const Link = forwardRef<React.ComponentRef<'a'>, LinkProps>(
  ({ mode = 'internal', className, disabled, ...props }, ref) => (
    <a
      ref={ref}
      className={cn(linkStyles({ mode, className }), {
        'text-info-50/25 pointer-events-none': disabled
      })}
      {...props}
    />
  )
);
