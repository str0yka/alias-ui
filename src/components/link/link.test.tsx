import { render, screen } from '@testing-library/react';

import { getTestProps } from '@/lib';

import { Link } from './link';

describe('Link Component', () => {
  it('should match snapshot', () => {
    const internallLink = render(<Link mode='internal'>link</Link>);
    const externalLink = render(<Link mode='external'>link</Link>);
    const disabledLink = render(<Link disabled>link</Link>);

    expect(internallLink.container).toMatchSnapshot();
    expect(externalLink.container).toMatchSnapshot();
    expect(disabledLink.container).toMatchSnapshot();
  });

  it('should render link with internal mode by default', () => {
    render(<Link {...getTestProps('link')}>link</Link>);

    const linkElement = screen.getByTestId('link');
    expect(linkElement).toHaveClass('active:underline');
  });

  it('should render link with external mode when mode prop is external', () => {
    render(
      <Link
        mode='external'
        {...getTestProps('link')}
      >
        link
      </Link>
    );

    const linkElement = screen.getByTestId('link');
    expect(linkElement).toHaveClass('underline', 'decoration-dashed', 'active:no-underline');
  });

  it('should render link with disabled classes', () => {
    render(
      <Link
        disabled
        {...getTestProps('link')}
      >
        link
      </Link>
    );

    const linkElement = screen.getByTestId('link');
    expect(linkElement).toHaveClass('text-info-50/25', 'pointer-events-none');
  });

  it('should render link with custom className', () => {
    render(
      <Link
        className='custom-class'
        {...getTestProps('link')}
      >
        link
      </Link>
    );

    const linkElement = screen.getByTestId('link');
    expect(linkElement).toHaveClass('custom-class');
  });
});
