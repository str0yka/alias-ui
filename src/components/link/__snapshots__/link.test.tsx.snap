// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link Component > should match snapshot 1`] = `
<div>
  <a
    class="typography-regular-12 text-info-400 cursor-pointer select-none underline-offset-2 hover:text-info-300 active:text-info-300 active:underline"
  >
    link
  </a>
</div>
`;

exports[`Link Component > should match snapshot 2`] = `
<div>
  <a
    class="typography-regular-12 text-info-400 cursor-pointer select-none underline-offset-2 hover:text-info-300 active:text-info-300 underline decoration-dashed active:no-underline"
  >
    link
  </a>
</div>
`;

exports[`Link Component > should match snapshot 3`] = `
<div>
  <a
    class="typography-regular-12 cursor-pointer select-none underline-offset-2 hover:text-info-300 active:text-info-300 active:underline text-info-50/25 pointer-events-none"
  >
    link
  </a>
</div>
`;
